File: rb/LasVegasCity_sdk.rb
# LasVegasCity SDK

require_relative 'utility/struct/voxgig_struct'
require_relative 'core/utility_type'
require_relative 'core/spec'
require_relative 'core/helpers'

# Load utility registration
require_relative 'utility/register'

# Load config and features
require_relative 'config'
require_relative 'feature/base_feature'
require_relative 'features'

# Load typed models (Struct value objects).
require_relative 'LasVegasCity_types'


class LasVegasCitySDK
  attr_accessor :mode, :features, :options

  def initialize(options = {})
    @mode = "live"
    @features = []
    @options = nil

    utility = LasVegasCityUtility.new
    @_utility = utility

    config = LasVegasCityConfig.make_config

    @_rootctx = utility.make_context.call({
      "client" => self,
      "utility" => utility,
      "config" => config,
      "options" => options || {},
      "shared" => {},
    }, nil)

    @options = utility.make_options.call(@_rootctx)

    if VoxgigStruct.getpath(@options, "feature.test.active") == true
      @mode = "test"
    end

    @_rootctx.options = @options

    # Add features in the resolved order (make_options puts an explicit array
    # order first, else defaults to test-first). Ordering matters: the `test`
    # feature installs the base mock transport and the transport features
    # (retry/cache/netsim/proxy/ratelimit) wrap whatever is current, so `test`
    # must be added before them to sit at the base of the chain.
    feature_opts = LasVegasCityHelpers.to_map(VoxgigStruct.getprop(@options, "feature"))
    if feature_opts
      featureorder = VoxgigStruct.getpath(@options, "__derived__.featureorder")
      if featureorder.is_a?(Array)
        featureorder.each do |fname|
          fopts = LasVegasCityHelpers.to_map(feature_opts[fname])
          if fopts && fopts["active"] == true
            utility.feature_add.call(@_rootctx, LasVegasCityFeatures.make_feature(fname))
          end
        end
      end
    end

    # Add extension features.
    extend_val = VoxgigStruct.getprop(@options, "extend")
    if extend_val.is_a?(Array)
      extend_val.each do |f|
        if f.respond_to?(:get_name)
          utility.feature_add.call(@_rootctx, f)
        end
      end
    end

    # Initialize features.
    @features.each do |f|
      utility.feature_init.call(@_rootctx, f)
    end

    utility.feature_hook.call(@_rootctx, "PostConstruct")
  end

  def options_map
    out = VoxgigStruct.clone(@options)
    out.is_a?(Hash) ? out : {}
  end

  def get_utility
    LasVegasCityUtility.copy(@_utility)
  end

  def get_root_ctx
    @_rootctx
  end

  def prepare(fetchargs = {})
    utility = @_utility
    fetchargs ||= {}

    ctrl = LasVegasCityHelpers.to_map(VoxgigStruct.getprop(fetchargs, "ctrl")) || {}

    ctx = utility.make_context.call({
      "opname" => "prepare",
      "ctrl" => ctrl,
    }, @_rootctx)

    opts = @options
    path = VoxgigStruct.getprop(fetchargs, "path") || ""
    path = "" unless path.is_a?(String)
    method_val = VoxgigStruct.getprop(fetchargs, "method") || "GET"
    method_val = "GET" unless method_val.is_a?(String)
    params = LasVegasCityHelpers.to_map(VoxgigStruct.getprop(fetchargs, "params")) || {}
    query = LasVegasCityHelpers.to_map(VoxgigStruct.getprop(fetchargs, "query")) || {}
    headers = utility.prepare_headers.call(ctx)

    base = VoxgigStruct.getprop(opts, "base") || ""
    base = "" unless base.is_a?(String)
    prefix = VoxgigStruct.getprop(opts, "prefix") || ""
    prefix = "" unless prefix.is_a?(String)
    suffix = VoxgigStruct.getprop(opts, "suffix") || ""
    suffix = "" unless suffix.is_a?(String)

    ctx.spec = LasVegasCitySpec.new({
      "base" => base, "prefix" => prefix, "suffix" => suffix,
      "path" => path, "method" => method_val,
      "params" => params, "query" => query, "headers" => headers,
      "body" => VoxgigStruct.getprop(fetchargs, "body"),
      "step" => "start",
    })

    # Merge user-provided headers.
    uh = VoxgigStruct.getprop(fetchargs, "headers")
    if uh.is_a?(Hash)
      uh.each { |k, v| ctx.spec.headers[k] = v }
    end

    _, err = utility.prepare_auth.call(ctx)
    raise err if err

    # make_fetch_def returns a (fetchdef, err) tuple; destructure it and
    # return just the fetchdef Hash (raising on error) so callers — including
    # direct(), which indexes fetchdef["url"] — receive a Hash, mirroring the
    # ts/py prepare().
    fetchdef, fd_err = utility.make_fetch_def.call(ctx)
    raise fd_err if fd_err

    fetchdef
  end

  def direct(fetchargs = {})
    utility = @_utility

    # direct() is the raw-HTTP escape hatch: it always returns a result hash
    # ({ "ok" => ..., ... }) and never raises. prepare() raises on error, so
    # trap that and surface it in the hash.
    begin
      fetchdef = prepare(fetchargs)
    rescue LasVegasCityError => err
      return { "ok" => false, "err" => err }
    end

    fetchargs ||= {}
    ctrl = LasVegasCityHelpers.to_map(VoxgigStruct.getprop(fetchargs, "ctrl")) || {}

    ctx = utility.make_context.call({
      "opname" => "direct",
      "ctrl" => ctrl,
    }, @_rootctx)

    url = fetchdef["url"] || ""
    fetched, fetch_err = utility.fetcher.call(ctx, url, fetchdef)

    return { "ok" => false, "err" => fetch_err } if fetch_err

    if fetched.nil?
      return {
        "ok" => false,
        "err" => ctx.make_error("direct_no_response", "response: undefined"),
      }
    end

    if fetched.is_a?(Hash)
      status = LasVegasCityHelpers.to_int(VoxgigStruct.getprop(fetched, "status"))
      headers = VoxgigStruct.getprop(fetched, "headers") || {}

      # No-body responses (204, 304) and explicit zero content-length must
      # skip JSON parsing — calling json() on an empty body errors.
      content_length = headers.is_a?(Hash) ? headers["content-length"] : nil
      no_body = status == 204 || status == 304 || content_length.to_s == "0"

      json_data = nil
      unless no_body
        jf = VoxgigStruct.getprop(fetched, "json")
        if jf.is_a?(Proc)
          begin
            json_data = jf.call
          rescue StandardError
            # Non-JSON body — leave data nil, keep status/headers.
            json_data = nil
          end
        end
      end

      return {
        "ok" => status >= 200 && status < 300,
        "status" => status,
        "headers" => headers,
        "data" => json_data,
      }
    end

    return {
      "ok" => false,
      "err" => ctx.make_error("direct_invalid", "invalid response type"),
    }
  end


  # Canonical facade: client.CityInfo.list / client.CityInfo.load({ "id" => ... })
  def CityInfo(data = nil)
    require_relative 'entity/city_info_entity'
    CityInfoEntity.new(self, data)
  end


  # Canonical facade: client.Council.list / client.Council.load({ "id" => ... })
  def Council(data = nil)
    require_relative 'entity/council_entity'
    CouncilEntity.new(self, data)
  end


  # Canonical facade: client.Department.list / client.Department.load({ "id" => ... })
  def Department(data = nil)
    require_relative 'entity/department_entity'
    DepartmentEntity.new(self, data)
  end


  # Canonical facade: client.EconomicDevelopment.list / client.EconomicDevelopment.load({ "id" => ... })
  def EconomicDevelopment(data = nil)
    require_relative 'entity/economic_development_entity'
    EconomicDevelopmentEntity.new(self, data)
  end


  # Canonical facade: client.Event.list / client.Event.load({ "id" => ... })
  def Event(data = nil)
    require_relative 'entity/event_entity'
    EventEntity.new(self, data)
  end


  # Canonical facade: client.Job.list / client.Job.load({ "id" => ... })
  def Job(data = nil)
    require_relative 'entity/job_entity'
    JobEntity.new(self, data)
  end


  # Canonical facade: client.Meeting.list / client.Meeting.load({ "id" => ... })
  def Meeting(data = nil)
    require_relative 'entity/meeting_entity'
    MeetingEntity.new(self, data)
  end


  # Canonical facade: client.New.list / client.New.load({ "id" => ... })
  def New(data = nil)
    require_relative 'entity/new_entity'
    NewEntity.new(self, data)
  end


  # Canonical facade: client.Park.list / client.Park.load({ "id" => ... })
  def Park(data = nil)
    require_relative 'entity/park_entity'
    ParkEntity.new(self, data)
  end


  # Canonical facade: client.Permit.list / client.Permit.load({ "id" => ... })
  def Permit(data = nil)
    require_relative 'entity/permit_entity'
    PermitEntity.new(self, data)
  end


  # Canonical facade: client.PublicSafety.list / client.PublicSafety.load({ "id" => ... })
  def PublicSafety(data = nil)
    require_relative 'entity/public_safety_entity'
    PublicSafetyEntity.new(self, data)
  end



  def self.test(testopts = nil, sdkopts = nil)
    sdkopts = sdkopts || {}
    sdkopts = VoxgigStruct.clone(sdkopts)
    sdkopts = {} unless sdkopts.is_a?(Hash)

    testopts = testopts || {}
    testopts = VoxgigStruct.clone(testopts)
    testopts = {} unless testopts.is_a?(Hash)
    testopts["active"] = true

    VoxgigStruct.setpath(sdkopts, "feature.test", testopts)

    sdk = LasVegasCitySDK.new(sdkopts)
    sdk.mode = "test"
    sdk
  end
end

File: py/lasvegascity_sdk.py
# LasVegasCity SDK

from utility.voxgig_struct import voxgig_struct as vs
from core.utility_type import LasVegasCityUtility
from core.spec import LasVegasCitySpec
from core import helpers

# Load utility registration (populates Utility._registrar)
from utility import register

# Load features
from feature.base_feature import LasVegasCityBaseFeature
from features import _make_feature


class LasVegasCitySDK:

    def __init__(self, options=None):
        self.mode = "live"
        self.features = []
        self.options = None

        utility = LasVegasCityUtility()
        self._utility = utility

        from config import make_config
        config = make_config()

        self._rootctx = utility.make_context({
            "client": self,
            "utility": utility,
            "config": config,
            "options": options if options is not None else {},
            "shared": {},
        }, None)

        self.options = utility.make_options(self._rootctx)

        if vs.getpath(self.options, "feature.test.active") is True:
            self.mode = "test"

        self._rootctx.options = self.options

        # Add features from config.
        feature_opts = helpers.to_map(vs.getprop(self.options, "feature"))
        if feature_opts is not None:
            feature_items = vs.items(feature_opts)
            if feature_items is not None:
                for item in feature_items:
                    fname = item[0]
                    fopts = helpers.to_map(item[1])
                    if fopts is not None and fopts.get("active") is True:
                        utility.feature_add(self._rootctx, _make_feature(fname))

        # Add extension features.
        extend = vs.getprop(self.options, "extend")
        if isinstance(extend, list):
            for f in extend:
                if isinstance(f, dict) or (hasattr(f, "get_name") and callable(f.get_name)):
                    utility.feature_add(self._rootctx, f)

        # Initialize features.
        for f in self.features:
            utility.feature_init(self._rootctx, f)

        utility.feature_hook(self._rootctx, "PostConstruct")

        # #BuildFeatures

    def options_map(self):
        out = vs.clone(self.options)
        if isinstance(out, dict):
            return out
        return {}

    def get_utility(self):
        return LasVegasCityUtility.copy(self._utility)

    def get_root_ctx(self):
        return self._rootctx

    def prepare(self, fetchargs=None):
        utility = self._utility

        if fetchargs is None:
            fetchargs = {}

        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "prepare",
            "ctrl": ctrl,
        }, self._rootctx)

        options = self.options

        path = vs.getprop(fetchargs, "path") or ""
        if not isinstance(path, str):
            path = ""

        method = vs.getprop(fetchargs, "method") or "GET"
        if not isinstance(method, str):
            method = "GET"

        params = helpers.to_map(vs.getprop(fetchargs, "params"))
        if params is None:
            params = {}
        query = helpers.to_map(vs.getprop(fetchargs, "query"))
        if query is None:
            query = {}

        headers = utility.prepare_headers(ctx)

        base = vs.getprop(options, "base") or ""
        if not isinstance(base, str):
            base = ""
        prefix = vs.getprop(options, "prefix") or ""
        if not isinstance(prefix, str):
            prefix = ""
        suffix = vs.getprop(options, "suffix") or ""
        if not isinstance(suffix, str):
            suffix = ""

        ctx.spec = LasVegasCitySpec({
            "base": base,
            "prefix": prefix,
            "suffix": suffix,
            "path": path,
            "method": method,
            "params": params,
            "query": query,
            "headers": headers,
            "body": vs.getprop(fetchargs, "body"),
            "step": "start",
        })

        # Merge user-provided headers.
        uh = vs.getprop(fetchargs, "headers")
        if isinstance(uh, dict):
            for k, v in uh.items():
                ctx.spec.headers[k] = v

        _, err = utility.prepare_auth(ctx)
        if err is not None:
            raise err

        fetchdef, err = utility.make_fetch_def(ctx)
        if err is not None:
            raise err

        return fetchdef

    def direct(self, fetchargs=None):
        utility = self._utility

        try:
            fetchdef = self.prepare(fetchargs)
        except Exception as err:
            # direct() is the raw-HTTP escape hatch: it never raises, it
            # returns a result object callers branch on via result["ok"].
            return {"ok": False, "err": err}

        if fetchargs is None:
            fetchargs = {}
        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "direct",
            "ctrl": ctrl,
        }, self._rootctx)

        url = fetchdef.get("url", "")
        fetched, fetch_err = utility.fetcher(ctx, url, fetchdef)

        if fetch_err is not None:
            return {"ok": False, "err": fetch_err}

        if fetched is None:
            return {
                "ok": False,
                "err": ctx.make_error("direct_no_response", "response: undefined"),
            }

        if isinstance(fetched, dict):
            status = helpers.to_int(vs.getprop(fetched, "status"))
            headers = vs.getprop(fetched, "headers") or {}

            # No-body responses (204, 304) and explicit zero content-length
            # must skip JSON parsing — calling json() on an empty body raises.
            content_length = None
            if isinstance(headers, dict):
                content_length = headers.get("content-length")
            no_body = status in (204, 304) or str(content_length) == "0"

            json_data = None
            if not no_body:
                jf = vs.getprop(fetched, "json")
                if callable(jf):
                    try:
                        json_data = jf()
                    except Exception:
                        # Non-JSON body (e.g. text/plain, text/html). Surface
                        # status + headers but leave data as None.
                        json_data = None

            return {
                "ok": status >= 200 and status < 300,
                "status": status,
                "headers": headers,
                "data": json_data,
            }

        return {
            "ok": False,
            "err": ctx.make_error("direct_invalid", "invalid response type"),
        }


    @property
    def city_info(self):
        """Idiomatic facade: client.city_info.list() / client.city_info.load({"id": ...})."""
        from entity.city_info_entity import CityInfoEntity
        cached = getattr(self, "_city_info", None)
        if cached is None:
            cached = CityInfoEntity(self, None)
            self._city_info = cached
        return cached

    def CityInfo(self, data=None):
        # Deprecated: use client.city_info instead.
        from entity.city_info_entity import CityInfoEntity
        return CityInfoEntity(self, data)


    @property
    def council(self):
        """Idiomatic facade: client.council.list() / client.council.load({"id": ...})."""
        from entity.council_entity import CouncilEntity
        cached = getattr(self, "_council", None)
        if cached is None:
            cached = CouncilEntity(self, None)
            self._council = cached
        return cached

    def Council(self, data=None):
        # Deprecated: use client.council instead.
        from entity.council_entity import CouncilEntity
        return CouncilEntity(self, data)


    @property
    def department(self):
        """Idiomatic facade: client.department.list() / client.department.load({"id": ...})."""
        from entity.department_entity import DepartmentEntity
        cached = getattr(self, "_department", None)
        if cached is None:
            cached = DepartmentEntity(self, None)
            self._department = cached
        return cached

    def Department(self, data=None):
        # Deprecated: use client.department instead.
        from entity.department_entity import DepartmentEntity
        return DepartmentEntity(self, data)


    @property
    def economic_development(self):
        """Idiomatic facade: client.economic_development.list() / client.economic_development.load({"id": ...})."""
        from entity.economic_development_entity import EconomicDevelopmentEntity
        cached = getattr(self, "_economic_development", None)
        if cached is None:
            cached = EconomicDevelopmentEntity(self, None)
            self._economic_development = cached
        return cached

    def EconomicDevelopment(self, data=None):
        # Deprecated: use client.economic_development instead.
        from entity.economic_development_entity import EconomicDevelopmentEntity
        return EconomicDevelopmentEntity(self, data)


    @property
    def event(self):
        """Idiomatic facade: client.event.list() / client.event.load({"id": ...})."""
        from entity.event_entity import EventEntity
        cached = getattr(self, "_event", None)
        if cached is None:
            cached = EventEntity(self, None)
            self._event = cached
        return cached

    def Event(self, data=None):
        # Deprecated: use client.event instead.
        from entity.event_entity import EventEntity
        return EventEntity(self, data)


    @property
    def job(self):
        """Idiomatic facade: client.job.list() / client.job.load({"id": ...})."""
        from entity.job_entity import JobEntity
        cached = getattr(self, "_job", None)
        if cached is None:
            cached = JobEntity(self, None)
            self._job = cached
        return cached

    def Job(self, data=None):
        # Deprecated: use client.job instead.
        from entity.job_entity import JobEntity
        return JobEntity(self, data)


    @property
    def meeting(self):
        """Idiomatic facade: client.meeting.list() / client.meeting.load({"id": ...})."""
        from entity.meeting_entity import MeetingEntity
        cached = getattr(self, "_meeting", None)
        if cached is None:
            cached = MeetingEntity(self, None)
            self._meeting = cached
        return cached

    def Meeting(self, data=None):
        # Deprecated: use client.meeting instead.
        from entity.meeting_entity import MeetingEntity
        return MeetingEntity(self, data)


    @property
    def new(self):
        """Idiomatic facade: client.new.list() / client.new.load({"id": ...})."""
        from entity.new_entity import NewEntity
        cached = getattr(self, "_new", None)
        if cached is None:
            cached = NewEntity(self, None)
            self._new = cached
        return cached

    def New(self, data=None):
        # Deprecated: use client.new instead.
        from entity.new_entity import NewEntity
        return NewEntity(self, data)


    @property
    def park(self):
        """Idiomatic facade: client.park.list() / client.park.load({"id": ...})."""
        from entity.park_entity import ParkEntity
        cached = getattr(self, "_park", None)
        if cached is None:
            cached = ParkEntity(self, None)
            self._park = cached
        return cached

    def Park(self, data=None):
        # Deprecated: use client.park instead.
        from entity.park_entity import ParkEntity
        return ParkEntity(self, data)


    @property
    def permit(self):
        """Idiomatic facade: client.permit.list() / client.permit.load({"id": ...})."""
        from entity.permit_entity import PermitEntity
        cached = getattr(self, "_permit", None)
        if cached is None:
            cached = PermitEntity(self, None)
            self._permit = cached
        return cached

    def Permit(self, data=None):
        # Deprecated: use client.permit instead.
        from entity.permit_entity import PermitEntity
        return PermitEntity(self, data)


    @property
    def public_safety(self):
        """Idiomatic facade: client.public_safety.list() / client.public_safety.load({"id": ...})."""
        from entity.public_safety_entity import PublicSafetyEntity
        cached = getattr(self, "_public_safety", None)
        if cached is None:
            cached = PublicSafetyEntity(self, None)
            self._public_safety = cached
        return cached

    def PublicSafety(self, data=None):
        # Deprecated: use client.public_safety instead.
        from entity.public_safety_entity import PublicSafetyEntity
        return PublicSafetyEntity(self, data)



    @classmethod
    def test(cls, testopts=None, sdkopts=None):
        if sdkopts is None:
            sdkopts = {}
        sdkopts = vs.clone(sdkopts)
        if not isinstance(sdkopts, dict):
            sdkopts = {}

        if testopts is None:
            testopts = {}
        testopts = vs.clone(testopts)
        if not isinstance(testopts, dict):
            testopts = {}
        testopts["active"] = True

        vs.setpath(sdkopts, "feature.test", testopts)

        sdk = cls(sdkopts)
        sdk.mode = "test"

        return sdk
